programs/mango-v4/src/instructions/perp_liq_base_and_positive_pnl.rs

```diff
@@ -0,0 +1,434 @@
+    );
+
+    let liqor_perp_position = liqor
+        .ensure_perp_position(perp_market_index, perp_market.settle_token_index)?
+        .0;
+
+    // Settle funding
+    liqee_perp_position.settle_funding(&perp_market);
+    liqor_perp_position.settle_funding(&perp_market);
+
+    // Max settleable on the liqee?
+    let now_ts: u64 = Clock::get()?.unix_timestamp.try_into().unwrap();
+    liqee_perp_position.update_settle_limit(&perp_market, now_ts);
+    let liqee_positive_settle_limit = liqee_perp_position.available_settle_limit(&perp_market).1;
+
+    // The max settleable amount does not need to be constrained by the liqor's perp settle health,
+    // because taking over perp pnl decreases liqor health: every unit of pnl taken costs
+    // (1-positive_pnl_liq_fee) USDC and only gains init_pnl_asset_weight in perp health.
+    let max_settle = I80F48::from(max_quote_transfer);
+
+    // Take over the liqee's base in exchange for quote
+    let liqee_base_lots = liqee_perp_position.base_position_lots();
+    // Each lot the base position gets closer to 0, the unweighted perp health contribution
+    // increases by this amount.
+    let unweighted_health_per_lot;
+    // -1 (liqee base lots decrease) or +1 (liqee base lots increase)
+    let direction: i64;
+    // Either 1+fee or 1-fee, depending on direction.
+    let fee_factor;
+    if liqee_base_lots > 0 {
+        require_msg!(
+            max_base_transfer >= 0,
+            "max_base_transfer can't be negative when liqee's base_position is positive"
+        );
+
+        // the unweighted perp health contribution gets reduced by `base * price * perp_init_asset_weight`
+        // and increased by `base * price * (1 - liq_fee) * quote_init_asset_weight`
+        let quote_init_asset_weight = I80F48::ONE;
+        direction = -1;
+        fee_factor = cm!(I80F48::ONE - perp_market.base_liquidation_fee);
+        unweighted_health_per_lot = cm!(price_per_lot
+            * (-perp_market.init_base_asset_weight + quote_init_asset_weight * fee_factor));
+    } else {
+        // liqee_base_lots <= 0
+        require_msg!(
+            max_base_transfer <= 0,
+            "max_base_transfer can't be positive when liqee's base_position is positive"
+        );
+
+        // health gets increased by `base * price * perp_init_liab_weight`
+        // and reduced by `base * price * (1 + liq_fee) * quote_init_liab_weight`
+        let quote_init_liab_weight = I80F48::ONE;
+        direction = 1;
+        fee_factor = cm!(I80F48::ONE + perp_market.base_liquidation_fee);
+        unweighted_health_per_lot = cm!(price_per_lot
+            * (perp_market.init_base_liab_weight - quote_init_liab_weight * fee_factor));
+    };
+    assert!(unweighted_health_per_lot > 0);
+
+    let spot_gain_per_settled = cm!(I80F48::ONE - perp_market.positive_pnl_liquidation_fee);
+    let init_overall_asset_weight = perp_market.init_pnl_asset_weight;
+    let expected_perp_health = |unweighted: I80F48| {
```

Comment:
nit: add comment that this is done here, instead of using the cache for cu saving?